### DISCUSSION TITLE
Multiple Audio Devices (UAC2) possible?

### DISCUSSION THREAD

**Discussion by mkarliner** [hathach/tinyusb#2026] [General]
I've been working on a device that is a composite UAC2 and Serial. This has worked OK, so I think I have the basics of device and interface descriptions OK.
I've been trying to change the device to present two UAC2 devices, and so far I've failed.

So, my first question is simple, does the UAC2 driver support multiple instances?
If it does, here is what I've done so far.

Running on Pi Pico and based on the UAC2 Headset example.

Changed tusb_config to

#define CFG_TUD_CDC               1
#define CFG_TUD_MSC               0
#define CFG_TUD_HID               0
#define CFG_TUD_MIDI              0
#define CFG_TUD_AUDIO             2
#define CFG_TUD_VENDOR            0

Changed usb_descriptors.c thus

```
uint8_t const desc_configuration[] =
    {
        // Interface count, string index, total length, attribute, power in mA
        TUD_CONFIG_DESCRIPTOR(1, ITF_NUM_TOTAL, 0, CONFIG_TOTAL_LEN, 0x00, 100),

        // Interface number, string index, EP Out & EP In address, EP size
        TUD_AUDIO_HEADSET_STEREO_DESCRIPTOR(2, EPNUM_AUDIO_OUT, EPNUM_AUDIO_IN | 0x80),
        //dual
      TUD_AUDIO_HEADSET_STEREO_DESCRIPTOR(4, EPNUM_AUDIO_OUT | 0x06, EPNUM_AUDIO_IN | 0x86),
        // Interface number, string index, EP notification address and size, EP data address (out, in) and size.
        TUD_CDC_DESCRIPTOR(ITF_NUM_CDC, 6, EPNUM_CDC_NOTIF, 8, EPNUM_CDC_OUT, EPNUM_CDC_IN, 64),
};
```

Change the enum in usb_descriptors.h thus:

```
enum
{
  ITF_NUM_AUDIO_CONTROL = 0,
  //dual
  ITF_NUM_AUDIO_STREAMING_SPK,
  ITF_NUM_AUDIO_STREAMING_MIC,
  ITF_NUM_AUDIO_STREAMING_SPK2,
  ITF_NUM_AUDIO_STREAMING_MIC2,
  ITF_NUM_CDC,
  ITF_CDC_DATA,
  ITF_NUM_TOTAL
};
```

This results in asserts of the form
process_set_config 903: ASSERT FAILED
process_control_request 706: ASSERT FAILED

which I believe are due to a failure to find a driver for one of the interfaces.

Any clues would be gratefully recieved. 
My apologies if I have not supplied enough or the wrong information.
similar apologies for hard coded integers and similar horrors.

Mike

**Comment by HiFiPhile** [comment]
You must have something wrong with your descriptor. Multiple device is possible, here is the configuration part of my composite device:

- A 32bit/384kHz output interface
- A 24bit/48kHz mic input interface
- A DFU interface
- A vendor class debug interface

[desc.zip](https://github.com/hathach/tinyusb/files/11248710/desc.zip)

**Reply by mkarliner** [reply]
Only just seen this (Github, why no notification?)
Thanks very, extremely, much.
I shall work on it.

**Reply by mkarliner** [reply]
Well, you were right about the descriptor being wrong.
After much headscratching, it turned out that the macro
TUD_AUDIO_HEADSET_STEREO_DESCRIPTOR(2, EPNUM_AUDIO_OUT, EPNUM_AUDIO_IN | 0x80),
Which I'd copied from the example, assumed fixed interface numbers.
This led to an attempt to reuse the same interface for the second device, which caused
a shower of ASSERTs.
I added a parameter for the first interface number and the composite device now comes up.
However, only the first audio device appears on my host system, in this case a Mac.
Any suggestions about where to look?  As far as I can see, the system is seeing the device and
setting clock rates etc.

BTW - Thanks very much for the pointer!

**Reply by HiFiPhile** [reply]
Have you tried on Windows ?

On Linux there is a catch, both devices' input/output terminal need to have the same number :
```c
// Unit numbers are arbitrary selected
// Output path
#define UAC2_ENTITY_USB2IIS_CLOCK           0x04
#define UAC2_ENTITY_USB2IIS_INPUT_TERMINAL  0x01
#define UAC2_ENTITY_USB2IIS_FEATURE_UNIT    0x02
#define UAC2_ENTITY_USB2IIS_OUTPUT_TERMINAL 0x03
// Input path
#define UAC2_ENTITY_MIC_CLOCK               0x05
#define UAC2_ENTITY_MIC_INPUT_TERMINAL      0x01
#define UAC2_ENTITY_MIC_FEATURE_UNIT        0x06
#define UAC2_ENTITY_MIC_OUTPUT_TERMINAL     0x03
```

I don't know about Mac.

**Reply by mkarliner** [reply]
Found it. It was more of the same issue. There were other places in the Headset descriptor that assumed just one audio device.

Next issue is it now looks like tud_audio_write doesn't give me a way of specifying the endpoint number that comes in with tud_audio_tx_done_pre_load_cb, so I can route incoming mic data to the correct USB device. However, I'll start a new discussion about this for the benefit of future generations.

Am I correct in assuming this is not an issue for you as you have just an input only device and an output only device?

Thanks again.

**Reply by HiFiPhile** [reply]
You can use tud_audio_n_write

**Reply by mkarliner** [reply]
You are my hero.
